spec/lib/models/git_repository_spec.rb

```diff
@@ -0,0 +1,128 @@
+      subject.clone
+    end
+  end
+
+  describe '#checkout' do
+    let(:ref) { 'test-ref' }
+
+    it 'calls run_command with correct parameters' do
+      allow(subject).to receive(:within_repo).and_yield
+
+      expect(subject).to receive(:run_command).with("git checkout --force '#{ref}' --")
+      subject.checkout(ref)
+    end
+
+    it 'runs a command inside the repo' do
+      expect(subject).to receive(:within_repo)
+      subject.checkout(ref)
+    end
+  end
+
+  describe '#within_repo' do
+    it 'executes the block within the correct directory' do
+      check = false
+      block = lambda do
+        check = true
+      end
+
+      allow(Dir).to receive(:chdir).with(path).and_yield
+      subject.within_repo &block
+
+      expect(check).to be_truthy
+    end
+  end
+
+  describe '#add' do
+    let(:add_ref) { 'test-ref' }
+
+    it 'calls run_command with correct parameters' do
+      allow(subject).to receive(:within_repo).and_yield
+      expect(subject).to receive(:run_command).with("git add #{add_ref}")
+      subject.add(add_ref)
+    end
+  end
+
+  describe '#commit' do
+    let(:commit_message) { 'test-message' }
+
+    it 'calls run_command with correct parameters' do
+      allow(subject).to receive(:within_repo).and_yield
+      expect(subject).to receive(:run_command).with("git commit -m '#{commit_message}'")
+      subject.commit(commit_message)
+    end
+  end
+
+  describe '#push' do
+    it 'calls run_command with correct parameters' do
+      allow(subject).to receive(:within_repo).and_yield
+      expect(subject).to receive(:run_command).with("git push")
+      subject.push
+    end
+
+    context 'with pushing tags' do
+      it 'calls run_command with correct parameters' do
+        allow(subject).to receive(:within_repo).and_yield
+        expect(subject).to receive(:run_command).with("git push --tags")
+        subject.push(true)
+      end
+    end
+  end
+
+  describe '#branch' do
+    let(:branch) { 'test-branch' }
+    it 'returns the correct branch' do
+      allow(subject).to receive(:within_repo).and_yield
+      allow(subject).to receive(:run_command).and_return(branch)
+      expect(subject.branch).to eql(branch)
+    end
+  end
+
+  describe '#merge' do
+    let(:our_branch) { 'test-our-branch' }
+    let(:their_branch) { 'test-their-branch' }
+
+    before do
+      allow(subject).to receive(:within_repo).and_yield
+      allow(subject).to receive(:branch).and_return(our_branch)
+    end
+
+    it 'calls run_command with correct parameters' do
+      expected_command = "git merge origin/#{their_branch}"
+      expected_command += " -m \"Merge branch '#{their_branch}' into #{our_branch}'\""
+      expect(subject).to receive(:run_command).with(expected_command)
+      subject.merge(their_branch)
+    end
+
+    context 'strategy ours' do
+      it 'calls run_command with correct parameters' do
+        expected_command = "git merge origin/#{their_branch}"
+        expected_command += ' -Xours'
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.